### ISSUE TITLE
Support customizable short flags

### ISSUE THREAD

**Issue by evanunderscore** [evanunderscore/defopt#2]
Currently defopt allows for no customization of the command line. There are some things that could be made more flexible without violating the spirit of defopt if implemented well.

Possibilities:
* Short flags, e.g. `-c/--count`
* "store_const" flags, e.g. `--thing/--not-thing`
* Others?

One potential solution is to parse this information out of RST comments so the generated documentation is not polluted.

```
def func(count=1, up=False):
    """Do something

    :param int count: desc
    :param bool up: desc

    .. count: -c/--count COUNT
    .. up: --up=True --down=False
    """
```

**Comment by evanunderscore** [comment]
These comments are valid for Google/Numpy docstrings but must be above the fields.

**Comment by evanunderscore** [comment]
Going to hold off on implementing anything until I have a better idea of potential use cases.

**Comment by evanunderscore** [comment]
One downside doing this inside the docstring is that it introduces magic that doesn't exist outside defopt, which I've so far avoided. Perhaps it would be better to just pass something to `defopt.run`?

**Comment by flying-sheep** [comment]
or provide a decorator after all. my dream set of extensions:

* short opts
* simple way to use dashes instead of underscores
* flags with automatic `--no-<name>` and `+f` versions (or are they already there?)

```py
#!/usr/bin/env python
@defopt.short_opts(long='l', legend='g')
def my_cmd(long_opt=1, legend=True):
	print(long, legend)

if __name__ == '__main__':
	defopt.run(my_cmd, use_dashes=True)
```

```console
$ ./my-cmd --long-opt 5 -g
5 True
$ ./my-cmd +g
1 False
$ ./my-cmd --no-legend
1 False
```

**Comment by evanunderscore** [comment]
On dashes versus underscores: I hadn't considered this originally, but I feel like this would be good for the default behavior. If someone actually wants underscores in their flags, it could probably be an argument to `defopt.run`.

For decorators, this sort of thing is already quite nicely covered by [argh](http://argh.readthedocs.org/en/latest/reference.html#argh.decorators.arg), and also by [plac](http://plac.googlecode.com/hg/doc/plac.html#more-features) which I mentioned in the readme (I might put argh in there instead). Part of the reason I didn't want to include a decorator interface is that I don't want defopt to just become a less flexible equivalent of these existing parsers.

My thinking lately is that it would be nice to handle this in a similar way to parsers. In the same way the parsers argument to `defopt.run` maps types to parsers globally, this could map flag names to short flags globally. This is less flexible, but very simple and consistent.

The rough equivalent for your example:

```
def my_cmd(long_opt=1, legend=True, custom='a'):
    print(long_opt, legend, custom)

if __name__ == '__main__':
    defopt.run(my_cmd, short={'long_opt': 'l', 'legend': 'g', 'custom': 'c'})
```

How would you feel about something like this?

**Comment by evanunderscore** [comment]
I've renamed this issue so that this can focus on short flags and #12 can handle boolean flags.

**Comment by flying-sheep** [comment]
> My thinking lately is that it would be nice to handle this in a similar way to parsers. In the same way the parsers argument to defopt.run maps types to parsers globally, this could map flag names to short flags globally. This is less flexible, but very simple and consistent.
> […]
> ```py
> defopt.run(my_cmd, short={'long_opt': 'l', 'legend': 'g', 'custom': 'c'})
> ```

i like it, but we should consider doing `alternatives={'long_opt': ('l', 'long', 'long2'), 'legend': 'g'}` instead. it’s a more flexible than `short`. the behavior would be to

1. only have a long opt (from arg name) per default
2. if (one or more) short opts are given, use the default long opt and them
3. if (one or more) long opts are given, replace the default long opt

**Comment by evanunderscore** [comment]
Can you give me an example of where you would want to rename a flag but not just change the name of the function parameter?

**Comment by flying-sheep** [comment]
e.g. a backwards compatible flag name change?

it doesn’t matter though, as adding this part later would still work (i.e. implementing like you propose and later add the ability to handle tuples)

**Comment by evanunderscore** [comment]
My preference would be to implement a simpler solution first and expand on it later if necessary. I'm open to implementing something like you proposed in the future if there's a demand for it.

**Comment by flying-sheep** [comment]
yeah, sure :smile:

**Comment by evanunderscore** [comment]
Very simple implementation here: 94782990d9a0edc16602eb6c9c004d6d65ad46bd

Might have to change a bit once #12 is done. Thoughts?

**Comment by evanunderscore** [comment]
I've made the change from underscores to hyphens here: 75c87553efd310ed1f233854a2e9cac80dc6f778

**Comment by flying-sheep** [comment]
cool!

**Comment by evanunderscore** [comment]
Indeed, all of these changes will be in 2.0. The short flags could go into 1.x, but unless someone asks specifically, I won't bother.

**Comment by evanunderscore** [comment]
Implemented in c44e3980f581f508eaeaf9092294d05463fc2895 taking into account the other changes that have happened.

**Comment by evanunderscore** [comment]
Going to close this since most of it has been addressed. If there's anything you want that the current solution doesn't cover, feel free to open a new ticket for it.

**Comment by flying-sheep** [comment]
very cool, thank you!
